test/GPv2Transfer/TransferFromAccounts.t.sol

```diff
@@ -0,0 +1,171 @@
+// SPDX-License-Identifier: LGPL-3.0-or-later
+pragma solidity ^0.8;
+
+import {GPv2Order, GPv2Transfer, IERC20, IVault} from "src/contracts/libraries/GPv2Transfer.sol";
+
+import {Helper} from "./Helper.sol";
+
+contract TransferFromAccounts is Helper {
+    address private trader = makeAddr("GPv2Transfer.transferFromAccounts trader");
+    address payable private recipient = payable(makeAddr("GPv2Transfer.transferFromAccounts recipient"));
+    uint256 private amount = 0.1337 ether;
+
+    function test_should_transfer_ERC20_amount_to_recipient() public {
+        vm.mockCall(address(token), abi.encodeCall(IERC20.transferFrom, (trader, recipient, amount)), abi.encode(true));
+        GPv2Transfer.Data[] memory transfers = new GPv2Transfer.Data[](1);
+        transfers[0] =
+            GPv2Transfer.Data({account: trader, token: token, amount: amount, balance: GPv2Order.BALANCE_ERC20});
+        executor.transferFromAccountsTest(vault, transfers, recipient);
+    }
+
+    function test_should_transfer_external_amount_to_recipient() public {
+        IVault.UserBalanceOp[] memory vaultOps = new IVault.UserBalanceOp[](1);
+        vaultOps[0] = IVault.UserBalanceOp({
+            kind: IVault.UserBalanceOpKind.TRANSFER_EXTERNAL,
+            asset: token,
+            amount: amount,
+            sender: trader,
+            recipient: recipient
+        });
+        vm.mockCall(address(vault), abi.encodeCall(IVault.manageUserBalance, (vaultOps)), hex"");
+
+        GPv2Transfer.Data[] memory transfers = new GPv2Transfer.Data[](1);
+        transfers[0] =
+            GPv2Transfer.Data({account: trader, token: token, amount: amount, balance: GPv2Order.BALANCE_EXTERNAL});
+        executor.transferFromAccountsTest(vault, transfers, recipient);
+    }
+
+    function test_should_transfer_internal_amount_to_recipient() public {
+        IVault.UserBalanceOp[] memory vaultOps = new IVault.UserBalanceOp[](1);
+        vaultOps[0] = IVault.UserBalanceOp({
+            kind: IVault.UserBalanceOpKind.WITHDRAW_INTERNAL,
+            asset: token,
+            amount: amount,
+            sender: trader,
+            recipient: recipient
+        });
+        vm.mockCall(address(vault), abi.encodeCall(IVault.manageUserBalance, (vaultOps)), hex"");
+
+        GPv2Transfer.Data[] memory transfers = new GPv2Transfer.Data[](1);
+        transfers[0] =
+            GPv2Transfer.Data({account: trader, token: token, amount: amount, balance: GPv2Order.BALANCE_INTERNAL});
+        executor.transferFromAccountsTest(vault, transfers, recipient);
+    }
+
+    function test_should_transfer_many_external_and_internal_amounts_to_recipient() public {
+        uint256 numTraders = 42;
+        GPv2Transfer.Data[] memory transfers = new GPv2Transfer.Data[](numTraders);
+        IVault.UserBalanceOp[] memory expectedVaultOps = new IVault.UserBalanceOp[](numTraders / 3 * 2 + numTraders % 3);
```

Comment:
IMO this isn't readible (nor is the original test). I understand that we want to mix many different types of transfer operations, and that similar is done in the original test, but I too would argue with that test that it is not understandable what is explicitly happening. Where do special numbers like the 3 come from? There are a lot of constants here that have an implicit meaning - that should be explicit.

Reply:
Are the new changes helping?